### PR TITLE
PowerShell Quickstart

### PR DESCRIPTION
**Pull request by potatoqualitee**
Related issue: https://github.com/github/docs/issues/283

### Why:

As a PowerShell developer, I find GitHub Actions extremely useful and very straightforward. PowerShell is built into each of the runner images, and is supported out of the box.  It has a strong CI/CD presence and it makes a ton of sense to include PowerShell in the quickstart.

Note that I used the `Building and testing Python` as the template for language and format for consistency.

Also, two other PRs must be merged for this PR to be complete:

https://github.com/actions/cache/pull/431
https://github.com/actions/starter-workflows/pull/661

### What's being changed:

Added an entire PowerShell / Pester quickstart

### Check off the following:
- [x] All of the tests are passing.
- [?] I have reviewed my changes in staging.
- [x] For content changes, I have reviewed the [localization checklist](https://github.com/github/docs/blob/main/contributing/localization-checklist.md)
- [x] For content changes, I have reviewed the [Content style guide for GitHub Docs](https://github.com/github/docs/blob/main/contributing/content-style-guide.md).

Not sure what `reviewed my changes in staging` means, but I did look at it in my repo. Perhaps a link can be added to the issue template?

### CODE REVIEW

**Comment by janiceilene** [comment]
Thanks @potatoqualitee 💝 We'll get this over to the @github/docs-content-ecosystem to review!

**Comment by potatoqualitee** [comment]
Note that I recommended a PSScriptAnalyzer Action in the comments. The Pester project just reached out and suggested we recommend their gorgeous action as well, I'd love to! 

https://github.com/marketplace/actions/publish-unit-test-results

Is this something I can throw into the comments or expand upon? I didn't know about their action before and now I'm going to update my own projects.

File: content/actions/guides/building-and-testing-powershell.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    shell: pwsh
+    run: |
+      Set-PSRepository PSGallery -InstallationPolicy Trusted
+      Install-Module SqlServer, PSScriptAnalyzer -ErrorAction Stop
+```
+{% endraw %}
+
+### Testing your code
+
+You can use the same commands that you use locally to build and test your code.
```
Comment (rachmari) [inline content/actions/guides/building-and-testing-powershell.md:126]: What do you think about explaining this code in this section:

```
- name: Perform a Pester test from the command-line to ensure expected results
      shell: pwsh
      run: Get-ChildItem | Select-Object -ExpandProperty Name -First 1 | Should -Be 'bin'
    - name: Perform advanced tests
      shell: pwsh
      run: |
        Invoke-Pester Unit.Tests.ps1 -Passthru
        Invoke-Pester Integration.Tests.ps1 -Passthru
```

In the other guides, we use this section to talk about the test steps and break down any commands that might need more description.

File: content/actions/guides/building-and-testing-powershell.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  free-pro-team: '*'
+  enterprise-server: '>=2.22'
+---
+
+{% data reusables.actions.enterprise-beta %}
+{% data reusables.actions.enterprise-github-hosted-runners %}
+
+### Introduction
+
+This guide shows you how to install, test and publish a PowerShell module.
```
Comment (rachmari) [inline content/actions/guides/building-and-testing-powershell.md:14]: > install, test and publish a PowerShell module.

Does install here refer to installing dependencies?

File: content/actions/guides/building-and-testing-powershell.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
+---
+title: Building and testing PowerShell
+intro: You can create a continuous integration (CI) workflow to build and test your PowerShell project.
+versions:
```
Comment (rachmari) [inline content/actions/guides/building-and-testing-powershell.md:4]: This is something that was also missed in the Python guide, but we're updating that in another PR. So I wouldn't expect you to have included it here. 

It just displays this banner, which you'll see in the other actions pages:
Screen Shot 2020-10-09 at 4 11 41 PM




```suggestion
product: '{% data reusables.gated-features.actions %}'
versions:
```

**Review by rachmari — COMMENTED** [review]
@potatoqualitee wow!!! This is A-MAZ-ING! 👏 👏 👏 

Thank you so much for putting this together and adding the additional PRs for the caching example and starter template. I will preface this review with the fact that I have NEVER used PowerShell, and I'll reach out to see if someone with more PowerShell experience can also take a look. 

You did an excellent job emulating our other language guides when putting this together. Thank you for that! I had a few high-level questions about how the template relates to other sections of the guide. With some of those questions answered, I can suggest some more specific changes. 

Thanks again for putting together such a thorough guide. This will be great for the GitHub community! ❤️ :octocat:

File: content/actions/guides/building-and-testing-powershell.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+The table below describes the locations for various PowerShell modules in each {% data variables.product.prodname_dotcom %}-hosted runner.
+
+|| Ubuntu | Mac | Windows |
+|------|-------|------|----------|
+|**PowerShell system modules** |`/opt/microsoft/powershell/7/Modules/*`|`/usr/local/microsoft/powershell/7/Modules/*`|`C:\program files\powershell\7\Modules\*`|
+|**PowerShell add-on modules**|`/usr/local/share/powershell/Modules/*`|`/usr/local/share/powershell/Modules/*`|`C:\Modules\*`|
+|**User-installed modules**|`/home/runner/.local/share/powershell/Modules/*`|`/Users/runner/.local/share/powershell/Modules/*`|`C:\Users\runneradmin\Documents\PowerShell\Modules\*`|
+
+### Installing dependencies
```
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:75]: Thanks for the review! 

> I'm curious if there is anything in the Pester framework that allows you to install the dependencies in a more automated way

`Install-Module` is a core PowerShell command while `Pester` is the testing framework. `Install-Module` takes care of installing dependencies for modules when they come from the PowerShell Gallery but in this case, someone is testing their module *before* it goes into the Gallery, so the dependencies must be installed manually.

Install-Module accepts Name but not Path. It is not common PowerShell practice to get the list from a file.

> When developing PowerShell modules, is it common to install dependencies?

About 90% of my PowerShell modules have dependencies. Some developers may do it more than others, but all-in-all, it is a totally common practice. Common enough for the workflow template? Hmmm, I just did a quick poll on Twitter, I'll see what the community says 😁 https://twitter.com/cl/status/1314834653531049984

**Review by Diabolos720 — APPROVED** [review]

File: content/actions/guides/building-and-testing-powershell.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  build:
+    name: Pester tests work on all platforms
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macOS-latest]
+
+    steps:
+    - uses: actions/checkout@v1
+    - name: Perform a Pester test from the command-line to ensure expected results
```
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:55]: Pester testing is indeed part of a typical PowerShell CI workflow. The test is useless but provides a valid and simplified example.

Generally, we'll test from a file, and not right from the command line. I've done it at the command line before, then moved it to file. 

What I like about this example is that it shows PowerShell being used at the command line w/o being executed from a file. I do execute a lot of PS right from `run:` - just not as much for Pester testing. I hope that's clear 🤞

File: content/actions/guides/building-and-testing-powershell.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+{% data reusables.actions.enterprise-setup-prereq %}
+
+### Starting with the PowerShell workflow template
+
+{% data variables.product.prodname_dotcom %} provides a PowerShell workflow template that should work for most PowerShell projects. This guide includes examples that you can use to customize the template. For more information, see the [PowerShell workflow template](https://github.com/actions/starter-workflows/blob/main/ci/powershell.yml).
+
+To get started quickly, add the template to the `.github/workflows` directory of your repository.
+
+{% raw %}
+```yaml
```
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:35]: PowerShell generally is not built when people are creating modules using functions (C# cmdlets is a different story, but that's only a small segment of the PowerShell population). The way that modules are loaded is straightforward - files are loaded from the module directory.

We do use "builds" to clean up module directories before we publish them to the PowerShell Gallery (to remove directories like `.github`. Perhaps that is a good idea for the Publish section.

File: content/actions/guides/building-and-testing-powershell.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+
+We recommend that you have a basic understanding of PowerShell and Pester. For more information, see:
+- [Getting started with PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/learn/ps101/01-getting-started)
+- [Pester](https://pester.dev)
+
+{% data reusables.actions.enterprise-setup-prereq %}
+
+### Starting with the PowerShell workflow template
+
+{% data variables.product.prodname_dotcom %} provides a PowerShell workflow template that should work for most PowerShell projects. This guide includes examples that you can use to customize the template. For more information, see the [PowerShell workflow template](https://github.com/actions/starter-workflows/blob/main/ci/powershell.yml).
```
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:30]: Is it me that updates the guide? I have a bad memory and worry this will forgotten.

File: content/actions/guides/building-and-testing-powershell.md (hunk 1 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:127]: That top one is such a simplified example, I'm not sure it's worth explaining - nobody will ever run this test. It is, however, the right structure for a test and is intended to give people ideas. 

Perhaps we remove the first test (that is unlikely to be used anyway) and just focus on the Unit and Integration tests. Those tests I wouldn't mind explaining. Do I do it in the comments?

File: content/actions/guides/building-and-testing-powershell.md (hunk 2 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:14]: great point! yes. perhaps it should be how to `install dependencies, test your module, and publish to the PowerShell Gallery`?

File: content/actions/guides/building-and-testing-powershell.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:77]: Looks like a majority of devs do have dependencies 😊

File: content/actions/guides/building-and-testing-powershell.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:56]: Thanks for all your great work here, @potatoqualitee ⚡ I also like having a command line example that demonstrates Pester. What do you think of an example that demonstrates how to use `test-path` from both the command line and the `Unit.Tests.ps1` file? With that approach, the reader can understand that there are two different ways of doing the same thing, but will appreciate that the Tests file is more scalable. I could also add in these updates if you like.

File: content/actions/guides/building-and-testing-powershell.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:31]: I agree that it makes sense to decouple the dependency for now. @potatoqualitee - would you mind if I revised this example into a standalone workflow?

File: content/actions/guides/building-and-testing-powershell.md (hunk 1 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:128]: I think the code explanations for this example could be in a bulleted list that follows after the code block, and would only need to be for the PowerShell-specific `run:` commands. For example: (assuming we use a `test-path` example as suggested in another comment)

- `run: Test-Path resultsfile.log | Should -Be $true` - Checks whether the file named `resultsfile.log` is present.
- `Invoke-Pester Unit.Tests.ps1 -Passthru` - Runs the tests defined in the `Unit.Tests.ps1` file.

File: content/actions/guides/building-and-testing-powershell.md (hunk 2 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:15]: That sounds good to me 👍

File: content/actions/guides/building-and-testing-powershell.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:56]: @martin389 sounds good. I had something similar in the original but the test was so silly, I ended up removing it. I'd love a better example. please add the updates 🚀

File: content/actions/guides/building-and-testing-powershell.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:31]: @martin389 please do 🙏

File: content/actions/guides/building-and-testing-powershell.md (hunk 1 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:128]: Please add this as well, if you can., I'm not sure I'm following exactly, @martin389

File: content/actions/guides/building-and-testing-powershell.md (hunk 2 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md]: Done, @rachmari ! 🚀

File: content/actions/guides/building-and-testing-powershell.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md]: Okay, great 🎉  I've added a commit with the changes 👍

File: content/actions/guides/building-and-testing-powershell.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md]: I've added a commit for this now 😄

File: content/actions/guides/building-and-testing-powershell.md (hunk 1 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:133]: 👍  Added the list of explanations, and an example screenshot and `Unit.Tests.ps1` entry.

**Comment by martin389** [comment]
Hi @potatoqualitee - I've added a commit with these changes: 
- Restructured example into a standalone procedure.
- Changed example to use `test-path`.
- Added explanations for the PowerShell actions:
  - Added screenshot of example test failure.
  - Added `Unit.Tests.ps1` example.

Would these work for you? 🙇‍♂️

**Comment by potatoqualitee** [comment]
Looks great, @martin389! I'm ready to merge when you all are 💯

**Comment by martin389** [comment]
> Looks great, @martin389! I'm ready to merge when you all are 💯

Thanks @potatoqualitee! I added some more edits now, and I think there was someone else who wanted to review this draft, so I'm just checking with them.

**Comment by martin389** [comment]
For folks reviewing this draft: a rendered preview is available here: 
https://docs-284--powershell.herokuapp.com/en/free-pro-team@latest/actions/guides/building-and-testing-powershell

**Comment by potatoqualitee** [comment]
super cool, can't wait!

File: content/actions/guides/building-and-testing-powershell.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+      run: |
+            Set-PSRepository PSGallery -InstallationPolicy Trusted
+            Install-Module PSScriptAnalyzer -ErrorAction Stop
+    - name: Lint with PSScriptAnalyzer
+      shell: pwsh
+      run: |
+            Invoke-ScriptAnalyzer -Path *.ps1 -Recurse -Outvariable issues
+            $errors   = $issues.Where({$_.Severity -eq 'Error'})
+            $warnings = $issues.Where({$_.Severity -eq 'Warning'})
+            Write-Output "There were $($errors.Count) errors and $($warnings.Count) warnings total."
```
Comment (martin389) [inline content/actions/guides/building-and-testing-powershell.md:161]: @potatoqualitee 
What do you think of using `write-error` or `throw` so that failures/warnings will cause the job to fail? Maybe something like this?
```suggestion
            If($warnings.Count -gt '0' -or $($errors.Count) -gt '0'){write-error "There were $($errors.Count) errors and $($warnings.Count) warnings total."}

```

Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:161]: Mine actually throw, too, not a bad idea. The only issue I have is with the syntax and capitalization of the proposed code. I'm also unsure if Write-Error throws an exception. To be sure, I'd add `-ErrorAction Stop`. I can't add a suggestion here, so I'll do it in the Files tab, brb

Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md]: I couldn't figure out how to suggest, so I added the following

````
if ($errors) {
    Write-Error "There were $($errors.Count) errors and $($warnings.Count) warnings total." -ErrorAction Stop
} else {
    Write-Output "There were $($errors.Count) errors and $($warnings.Count) warnings total."
}
````

Please let me know what you think. Warnings are common and can be looked over, so I just threw on error.

**Comment by martin389** [comment]
Hi @potatoqualitee 👋 
I have a status update for you here: The dev has completed their internal review, and mentioned a couple of things about the PowerShell module locations that I just need to confirm with another team. ⚡

File: content/actions/guides/building-and-testing-powershell.md [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+  Describe "Check results file is present" {
+      It "Check results file is present" {
+          Test-Path resultsfile.log | Should -Be $true
+      }
+  }
+  ```
+
+### PowerShell module locations
+
+The table below describes the locations for various PowerShell modules in each {% data variables.product.prodname_dotcom %}-hosted runner.
```
Comment (maxim-lobanov) [inline content/actions/guides/building-and-testing-powershell.md:74]: What is the reason to hardcode all theses locations in documentation?
1) All these paths are listed in `$env:PSModulePath` in powershell. Probably, it is better to rely on this variable.
2) I am not sure about cases where customers need these paths. The recommended use-case to install modules is `Install-Module` command that taking care about location by itself. I don't think that we should endorse any other ways to install modules

File: content/actions/guides/building-and-testing-powershell.md [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+|| Ubuntu | Mac | Windows |
+|------|-------|------|----------|
+|**PowerShell system modules** |`/opt/microsoft/powershell/7/Modules/*`|`/usr/local/microsoft/powershell/7/Modules/*`|`C:\program files\powershell\7\Modules\*`|
+|**PowerShell add-on modules**|`/usr/local/share/powershell/Modules/*`|`/usr/local/share/powershell/Modules/*`|`C:\Modules\*`|
+|**User-installed modules**|`/home/runner/.local/share/powershell/Modules/*`|`/Users/runner/.local/share/powershell/Modules/*`|`C:\Users\runneradmin\Documents\PowerShell\Modules\*`|
+
+### Installing dependencies
+
+{% data variables.product.prodname_dotcom %}-hosted runners have PowerShell 7 and Pester installed. You can use `Install-Module` to install additional dependencies from the PowerShell Gallery before building and testing your code. You can also cache dependencies to speed up your workflow. For more information, see "[Caching dependencies to speed up your workflow](/actions/automating-your-workflow-with-github-actions/caching-dependencies-to-speed-up-workflows)."
```
Comment (maxim-lobanov) [inline content/actions/guides/building-and-testing-powershell.md:84]: I think we should mention somewhere that we shouldn't rely on pre-installed packages.
Yes, Pester is pre-installed on every image but on images, we always install the latest version and update image every week. It can bring breaking changes.

For example, some time ago Pester was updated from 4.x to 5.x and brought a lot of breaking changes. Some customers who relied on pre-installed version, were broken.

I think it it worth to invoke `Install-Module` command always. If module is already pre-installed, this command will do nothing. I would even suggest for customers to use `Install-Module` with `-MaximumVersion` argument.

File: content/actions/guides/building-and-testing-powershell.md (hunk 8 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md]: @potatoqualitee - thanks so much, this looks good to me! 

Would it be useful to add the `ParseError` severity here? I saw that result come up in some of my Pester testing.

File: content/actions/guides/building-and-testing-powershell.md (hunk 9 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md:74]: I added them because they were in the Python docs, but also, I needed the information when creating my PowerShell module cacher. As a PowerShell Dev, I'd like having this documentation, as opposed to having to figure out my paths manually.

File: content/actions/guides/building-and-testing-powershell.md (hunk 10 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md:84]: Thanks for the review @maxim-lobanov 🎉  I'll work on getting these updates added.

File: content/actions/guides/building-and-testing-powershell.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+### Installing dependencies
+
+{% data variables.product.prodname_dotcom %}-hosted runners have PowerShell 7 and Pester installed. You can use `Install-Module` to install additional dependencies from the PowerShell Gallery before building and testing your code. 
+
+{% note %}
+
+**Note:** The Pester package used by {% data variables.product.prodname_dotcom %}-hosted runners is regularly updated, and can introduce signficant changes.
+As a result, it is recommended that you always specify the required package versions by using `Install-Module` with `-MaximumVersion`.
+
+{% endnote %}
```
Comment (martin389) [inline content/actions/guides/building-and-testing-powershell.md:91]: Hi @maxim-lobanov - does this note properly address what you had in mind?

Comment (maxim-lobanov) [reply content/actions/guides/building-and-testing-powershell.md:91]: Yep, that sounds good!
I would rephrase it a bit to generalize: `Pre-installed packages (like Pester) used by Hosted runners is regularly updated`

Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md]: Makes sense to me, thanks! I've added that in now.

**Review by maxim-lobanov — APPROVED** [review]

**Comment by martin389** [comment]
@potatoqualitee - I'm now done with adding updates from our internal reviewers, and this awesome guide looks ready to publish 🎉  

Now just busy with some final checks ⚡

File: content/actions/guides/building-and-testing-powershell.md (hunk 8 above)
Comment (potatoqualitee) [reply content/actions/guides/building-and-testing-powershell.md]: @martin389 - I don't know what a `ParseError` severity is, sorry. I don't use PSScriptAnalyzer outside of tests that people put in our repo 😅

**Comment by potatoqualitee** [comment]
@martin389 - you've been awesome! thanks for the assist 💙 and thank you for the review @maxim-lobanov

File: content/actions/guides/building-and-testing-powershell.md (hunk 8 above)
Comment (martin389) [reply content/actions/guides/building-and-testing-powershell.md]: No worries! I wasn't able to replicate it again in my testing, so it might have been a once-off event 😆

**Review by martin389 — APPROVED** [review]
Looks good to me 👍

**Comment by martin389** [comment]
Thanks @potatoqualitee! 🎊 
I'm expecting the guide to be available at the following link once all the automation completes:
https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-powershell

**Comment by martin389** [comment]
This guide is now published! 🎉  
https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-powershell

**Comment by potatoqualitee** [comment]
Awww yesss! Thanks again 🥳
